File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.py
class Solution:
    def minimumTime(self, time: List[int], totalTrips: int) -> int:
        left = min(time)
        right = min(time) * totalTrips

        while left < right:
            mid = left + (right - left) // 2
            if self.isPossible(mid, time, totalTrips):
                right = mid
            else:
                left = mid + 1

        return left

    def isPossible(self, mid: int, time: List[int], totalTrips: int) -> bool:
        total = 0
        for t in time:
            total += mid // t

        return total >= totalTrips

File: 2187-minimum-time-to-complete-trips/2187-minimum-time-to-complete-trips.java
class Solution {
    public long minimumTime(int[] time, int totalTrips) {
        long left = 1;
        long right = (long) 1e14;

        while (left < right) {
            long mid = left + (right - left) / 2;
            if (isPossible(mid, time, totalTrips)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }

        return left;
    }

    private boolean isPossible(long mid, int[] time, int totalTrips) {
        long total = 0;
        for (int t : time) {
            total += mid / t;
        }

        return total >= totalTrips;
    }
}
